share/completions/apm.fish

```diff
@@ -1,8 +1,227 @@
+function __fish_apm_needs_command
+    set cmd (commandline -opc)
+    if test (count $cmd) -eq 1
+        return 0
+    else if test (count $cmd) -eq 2
+        switch $cmd[2]
+            case '--color' '--no-color' 'help'
+                return 0
+            case '*'
+                return 1
+        end
+    end
+    return 1
+end
+
+function __fish_apm_using_command
+    set cmd (commandline -opc)
+    if test (count $cmd) -gt 1
+        for arg in $argv
+            if test $arg = $cmd[2]
+                return 0
+            end
+        end
+    end
+    return 1
+end
+
+function __fish_apm_config_args
+    if __fish_apm_using_command "config"
+        if test (count (commandline -opc)) -eq $argv
+            return 0
+        else
+            return 1
+        end
+    else
+        return 1
+    end
+end
+
+function __fish_apm_config_items
+    apm config list | grep -v '^;\|^$' | tr "\ =\ " "\t"
+end
+
+function __fish_apm_list_packages
+    apm list -b
+end
+
+if apm -h ^| grep -q "Atom Package Manager"
+    # Completions for Atom Package Manager
+
+    ##################
+    # global options #
+    ##################
+
+    # help
+    complete -f -c apm -n '__fish_apm_needs_command' -s h -l help -d "Display help"
+    complete -f -c apm -n '__fish_apm_needs_command' -a help -d "Display help"
+    complete -f -c apm -n '__fish_apm_using_command clean config dedupe deinstall delete dev develop docs erase featured home init install link linked links list ln lns login ls open outdated publish rebuild rebuild-module-cache remove rm search show star starred stars test uninstall unlink unpublish unstar update upgrade view' -s h -l help -d 'Display help'
+
+    # version
+    complete -f -c apm -n '__fish_apm_needs_command' -s v -l version -d "Display version"
+
+    # color
+    complete -f -c apm -n '__fish_apm_needs_command' -l color -d "Enable colored output"
+    complete -f -c apm -n '__fish_apm_needs_command' -l no-color -d "Disable colored output"
+
+    ############
+    # commands #
+    ############
+
+    # clean
+    complete -f -c apm -n '__fish_apm_needs_command' -a clean -d "Deletes packages in node_modules not referenced
+as dependency in package.json"
+
+    # config
+    complete -f -c apm -n '__fish_apm_needs_command' -a config -d "Modify or list configuration items"
+    complete -x -c apm -n '__fish_apm_config_args 2' -a set -d 'Set config item'
+    complete -x -c apm -n '__fish_apm_config_args 2' -a get -d 'Get config item'
+    complete -x -c apm -n '__fish_apm_config_args 2' -a delete -d 'Delete config item'
+    complete -x -c apm -n '__fish_apm_config_args 2' -a list -d 'List config items'
+    complete -x -c apm -n '__fish_apm_config_args 2' -a edit -d 'Edit config items'
+    complete -x -c apm -n '__fish_apm_config_args 3; and test (commandline -opc)[3] != "edit"; and test (commandline -opc)[3] != "list"' -a '(__fish_apm_config_items)'
```

Comment:
I would have probably split ...config_args into two functions - __fish_apm_config_{needs_command,_using_command_from}.

What you could do is `contains -- (commandline -opc)[3] edit list` instead of the tests. That'd be a tiny bit shorter and more easily extensible. (Of course this is all assuming that config has no options that can appear before the subcommand - if there are and you missed them or if it ever changes, this code will stop being correct)